components/merkledb/src/proof_list_index/proof_builder.rs

```diff
@@ -0,0 +1,119 @@
+//! Building `ListProof`s.
+
+use exonum_crypto::Hash;
+
+use std::ops::{Bound, RangeBounds};
+
+use super::{key::ProofListKey, tree_height_by_length, ListProof};
+use crate::BinaryValue;
+
+/// Encapsulation of a binary Merkle tree allowing to access its terminal and intermediate
+/// nodes.
+pub trait MerkleTree<V> {
+    /// Gets the length of the tree.
+    fn len(&self) -> u64;
+
+    /// Gets the node by its `position` in the tree.
+    ///
+    /// It is assumed that this method cannot fail since it is queried with `position`s
+    /// that are guaranteed to be present in the tree.
+    fn node(&self, position: ProofListKey) -> Hash;
+
+    /// Iterates over values starting from the specified index.
+    fn values<'s>(&'s self, start_index: u64) -> Box<dyn Iterator<Item = V> + 's>;
```

Comment:
Heap allocation could slow down things here, but AFAIK it's impossible to express the iterator type otherwise using stable Rust features (and even on nightly, [generic lifetimes for associated types seem to crash the compiler currently](https://github.com/rust-lang/rust/issues/62521)).